okhttp/src/main/java/okhttp3/internal/concurrent/TaskRunner.kt

```diff
@@ -27,12 +36,130 @@ package okhttp3.internal.concurrent
  *
  * Most applications should share a process-wide [TaskRunner] and use queues for per-client work.
  */
-interface TaskRunner {
-  fun newQueue(owner: Any): TaskQueue
+class TaskRunner(
+  val backend: Backend = RealBackend()
+) {
+  // All state in all tasks and queues is guarded by this.
+
+  private var coordinatorRunning = false
+  private val activeQueues = mutableListOf<TaskQueue>()
+  private val coordinator = Runnable { coordinate() }
+
+  fun newQueue(owner: Any) = TaskQueue(this, owner)
 
   /**
    * Returns a snapshot of queues that currently have tasks scheduled. The task runner does not
    * necessarily track queues that have no tasks scheduled.
    */
-  fun activeQueues(): List<TaskQueue>
+  fun activeQueues(): List<TaskQueue> {
+    synchronized(this) {
+      return activeQueues.toList()
+    }
+  }
+
+  internal fun kickCoordinator(queue: TaskQueue) {
+    check(Thread.holdsLock(this))
+
+    if (queue.isActive()) {
+      activeQueues.addIfAbsent(queue)
+    } else {
+      activeQueues.remove(queue)
+    }
+
+    if (coordinatorRunning) {
+      backend.coordinatorNotify(this)
+    } else {
+      coordinatorRunning = true
+      backend.executeCoordinator(coordinator)
+    }
+  }
+
+  private fun coordinate() {
+    synchronized(this) {
+      while (true) {
+        val now = backend.nanoTime()
+        val delayNanos = executeReadyTasks(now)
+
+        if (delayNanos == -1L) {
+          coordinatorRunning = false
+          return
+        }
+
+        try {
+          backend.coordinatorWait(this, delayNanos)
+        } catch (_: InterruptedException) {
+          // Will cause the coordinator to exit unless other tasks are scheduled!
+          cancelAll()
+        }
+      }
+    }
+  }
+
+  /**
+   * Start executing the next available tasks for all queues.
+   *
+   * Returns the delay until the next call to this method, -1L for no further calls, or
+   * [Long.MAX_VALUE] to wait indefinitely.
+   */
+  private fun executeReadyTasks(now: Long): Long {
+    var result = -1L
+
+    for (queue in activeQueues) {
+      val delayNanos = queue.executeReadyTask(now)
+      if (delayNanos == -1L) continue
+      result = if (result == -1L) delayNanos else minOf(result, delayNanos)
+    }
+
+    return result
+  }
+
+  private fun cancelAll() {
+    for (queue in activeQueues) {
+      queue.cancelAll()
+    }
+  }
+
+  interface Backend {
+    fun executeCoordinator(runnable: Runnable)
+    fun executeTask(runnable: Runnable)
+    fun nanoTime(): Long
+    fun coordinatorNotify(taskRunner: TaskRunner)
+    fun coordinatorWait(taskRunner: TaskRunner, nanos: Long)
+  }
+
+  class RealBackend : Backend {
+    private val coordinatorExecutor: Executor = ThreadPoolExecutor(
+        0, // corePoolSize.
+        1, // maximumPoolSize.
+        60L, TimeUnit.SECONDS, // keepAliveTime.
+        SynchronousQueue(),
+        threadFactory("OkHttp Task Coordinator", false)
```

Comment:
This seems like a good opportunity to throw the container users a bone and allow creating a backend with smaller timeouts or all daemon threads?  Can we make this simple e.g. a RealBackend factory method to create for different uses?

Reply:
My aspiration is to make container users happy by making the defaults work for them! Threads should promptly exit when there's no work to do.